internal/pubsub/rabbitpubsub/rabbit.go

```diff
@@ -0,0 +1,179 @@
+	pubc <-chan amqp.Confirmation // Go channel for server acks of publishes
+}
+
+// Values for the amqp client.
+// See https://www.rabbitmq.com/amqp-0-9-1-reference.html.
+const (
+	// Many methods of the amqp client take a "no-wait" parameter, which
+	// if true causes the client to return without waiting for a server
+	// response. We always want to wait.
+	wait = false
+
+	// Always use the empty routing key. This driver expects to be used with topic
+	// exchanges, which disregard the routing key.
+	routingKey = ""
+
+	// If the message can't be enqueued, return it to the sender rather than silently dropping it.
+	mandatory = true
+
+	// If there are no waiting consumers, enqueue the message instead of dropping it
+	immediate = true
+)
+
+// OpenTopic returns a *pubsub.Topic corresponding to the named exchange. The
+// exchange must have been previously created (for instance, by using
+// amqp.Channel.ExchangeDeclare). For the model of Go Cloud Pub/Sub to make sense,
+// the exchange should be a fanout exchange, although nothing in this package
+// enforces that.
+//
+// OpenTopic uses the supplied amqp.Connection for all communication. It is the caller's
+// responsibility to establish this connection before calling OpenTopic, and to close
+// it when it is no longer in use.
+//
+// The documentation of the amqp package recommends using separate connections for
+// publishing and subscribing.
+func OpenTopic(conn *amqp.Connection, name string) (*pubsub.Topic, error) {
+	// TODO(jba): support context.Context
+	t, err := newTopic(conn, name)
+	if err != nil {
+		return nil, err
+	}
+	return pubsub.NewTopic(t), nil
+
+}
+
+func newTopic(conn *amqp.Connection, name string) (*topic, error) {
+	t := &topic{
+		exchange: name,
+		conn:     conn,
+	}
+	if err := t.establishChannel(); err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+// establishChannel creates an AMQP channel if necessary. According to the amqp
+// package docs, once an error is returned from the channel, it must be discarded and
+// a new one created.
+//
+// Must be called with the t.mu held.
+func (t *topic) establishChannel() error {
+	// TODO(jba): support context.Context
+	if t.ch != nil {
+		// We already have a channel; nothing to do.
+		return nil
+	}
+	// Create a new channel.
+	ch, err := t.conn.Channel()
+	if err != nil {
+		return err
+	}
+	// Put the channel into a mode where confirmations are delivered for each publish.
+	if err := ch.Confirm(wait); err != nil {
+		return err
+	}
+	t.ch = ch
+	// Get a Go channel which will hold acks from the server. The server
+	// will send an ack for each published message to confirm that it was received.
+	t.pubc = ch.NotifyPublish(make(chan amqp.Confirmation)) // NotifyPublish returns its arg
+	return nil
+}
+
+// SendBatch implements driver.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
+	// It is simplest to allow only one SendBatch at a time. Allowing concurrent
```

Comment:
Outside this PR, but to me this is an indication that perhaps we should update `driver` to explicitly state that `SendBatch` will not be called concurrently, making driver implementations easier. At least we should be very explicit in the driver docstrings about whether this can happen and what the expectations are.

IIUC, we would enforce this by always setting the `max???` arg for the bundler to 1 (which it currently is), or getting rid of it.

Reply:
That would close off an important optimization. In cloud.google.com/go/pubsub, we make [many](https://github.com/GoogleCloudPlatform/google-cloud-go/blob/master/pubsub/topic.go#L439) concurrent publish RPCs to maximize throughput.